### ISSUE TITLE
Bug: Live recorded dem file sometimes throws "no unpacker" exception

### ISSUE THREAD

**Issue by ChitaAPC** [skadistats/clarity#236]
While running a game live, from a dem file which I record by using the `_record` Dota 2 console command I sometimes run into an exception. If an exception happens, it will consistently be thrown at the same time.

When running on my code, which uses clarity version 2.4.1 I get the following:
```
skadistats.clarity.processor.runner.ControllableRunner$4: Runner thread was terminated by an exception
	at skadistats.clarity.processor.runner.ControllableRunner.waitForTickReached(ControllableRunner.java:287)
	at skadistats.clarity.processor.runner.ControllableRunner.tick(ControllableRunner.java:321)
	at demofileparser.ParserRunner.run(ParserRunner.java:108)
	at java.lang.Thread.run(Thread.java:748)
Caused by: skadistats.clarity.ClarityException: entity at index 599 was not found for update.
	at skadistats.clarity.processor.entities.Entities.onPacketEntities(Entities.java:178)
	at skadistats.clarity.event.AbstractInvocationPoint.invoke(AbstractInvocationPoint.java:86)
	at skadistats.clarity.event.Event.raise(Event.java:37)
	at skadistats.clarity.processor.reader.InputSourceProcessor.processEmbedded(InputSourceProcessor.java:251)
	at skadistats.clarity.event.AbstractInvocationPoint.invoke(AbstractInvocationPoint.java:86)
	at skadistats.clarity.event.Event.raise(Event.java:37)
	at skadistats.clarity.processor.reader.InputSourceProcessor.processSource(InputSourceProcessor.java:178)
	at skadistats.clarity.event.AbstractInvocationPoint.invoke(AbstractInvocationPoint.java:86)
	at skadistats.clarity.event.Event.raise(Event.java:37)
	at skadistats.clarity.processor.runner.AbstractFileRunner.initAndRunWith(AbstractFileRunner.java:39)
	at skadistats.clarity.processor.runner.ControllableRunner$3.run(ControllableRunner.java:240)
	... 1 more
```
More context: 

- [here is one of the dem files](https://drive.google.com/file/d/1QU3WxPSVtmYs4mSbfIybmn_1CcKKiLv2/view?usp=sharing) which I am experiencing this.
- I am using a controllable runner and I reach this consistently when tick() is called at tick 80804 (bringing the tick counter to 80805)

When running clarity analyser with clarity version 2.5-SNAPSHOT:
```
skadistats.clarity.ClarityException: no unpacker for field m_hAbilities with type CHandle< CBaseEntity >[32]!
	at skadistats.clarity.decoder.s2.S2FieldReader.readFields(S2FieldReader.java:74) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.decoder.s2.S2FieldReader.readFields(S2FieldReader.java:14) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.processor.entities.Entities.processEntityUpdate(Entities.java:391) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.processor.entities.Entities.onPacketEntities(Entities.java:324) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.event.AbstractInvocationPoint.invoke(AbstractInvocationPoint.java:86) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.event.Event.raise(Event.java:37) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.processor.reader.InputSourceProcessor.processEmbedded(InputSourceProcessor.java:247) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.event.AbstractInvocationPoint.invoke(AbstractInvocationPoint.java:86) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.event.Event.raise(Event.java:37) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.processor.reader.InputSourceProcessor.processSource(InputSourceProcessor.java:174) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.event.AbstractInvocationPoint.invoke(AbstractInvocationPoint.java:86) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.event.Event.raise(Event.java:37) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.processor.runner.AbstractFileRunner.initAndRunWith(AbstractFileRunner.java:39) ~[clarity-2.5-SNAPSHOT.jar:na]
	at skadistats.clarity.processor.runner.ControllableRunner.lambda$runWith$0(ControllableRunner.java:238) ~[clarity-2.5-SNAPSHOT.jar:na]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_221]
```

I have spent a while trying to investigate this, it looks like entity 599 is in the list of updated entities, but I believe it was always null throughout the match. My guess is a bit got flipped during transmission to the dem file itself and we are pointing to the wrong entity.

My suggestion on this is to either catch the exception internally within `ControllableRunner.tick()` and allow the frame to be skipped entirely (or maybe catching it somewhere else like maybe `Entities.onPacketEntities` to enable just the one entity update to be skipped) or replace the exception raised with just a debug message.

I hope this helps, but if there is any more info I can give to try and fix this, please let me know

**Comment by spheenik** [comment]
Sorry for being so late: I checked the replay, and it seems to be corrupt: It tries to get an unpacker for a fixed array field base index, which is an invalid operation. This most certainly means that the replay data is corrupt shortly before.

Dota client replay writing has been wonky as of late.

Oh, and btw.: 2.4.1 is very bad at reading console recorded replays, so stick to 2.5-SNAPSHOT!

**Comment by nul0m** [comment]
i've faced exactly the same "no unpacker" issue using the latest clarity. @spheenik do you think there is any consistent way of skipping the failure and moving on to complete the parsing? @ChitaAPC did you try to work this around?

**Comment by spheenik** [comment]
You most certainly get this error in two cases:

1. normal replay is corrupted
2. console recorded replay

Console recorded replays contain data constellations I do not understand, so Clarity tries to work around them. This "no unpacker" issue is then a consequence of the workaround.

Best I can do at the moment is let you set a custom exception handler:

https://github.com/skadistats/clarity/blob/9ce7d1114a4cf81565af8c222f31c6714bb92d2e/src/main/java/skadistats/clarity/processor/runner/AbstractRunner.java#L69-L71

Call this on your runner before running and set your handler. The handler will get called when an exception occurs. If the handler does not throw, Clarity will continue parsing. Your mileage will vary. The state is not consistent afterwards. Some properties on some entities will be invalid.

**Comment by nul0m** [comment]
thanks a lot, @spheenik!! that's exactly what I was looking for!
maybe you know of any way how I could record the replay properly? i mean, i thought there are only two ways of retrieving the replay:
1. download from valve's server 
2. record it via console with the command `record` or `_record`

do I miss anything? if not, may there be a way to instruct the console's recorder to produce a proper dem file (e.g. a special option or console parameter)?

**Comment by spheenik** [comment]
Replays downloaded from Valve differ from console recorded replays. Unfortunately, there is no way to make the console recorded replay come out in the format that downloaded replays have - or at least I do not know such a way. I might have some improvements in the future though, they fix a special case for dota 2 console replays.

**Comment by nul0m** [comment]
great, thanks much!

**Comment by spheenik** [comment]
There should be some fixes from 2.6.0 onward, if there are still issues, please open another issue.
